thoth/storages/graph/postgres.py

```diff
@@ -775,16 +892,108 @@ def _construct_solved_query(
+
+        Edge cases:
+        CASE 1: ('package_name', None, 'index_url')
+
+        CASE 2: ('package_name', 'package_version', None)
+
+        CASE 3: ('package_name', None, None)
+        """
+        case_2 = (
+            self._session.query(PythonPackageVersionEntity)
+            .filter(
+                PythonPackageVersionEntity.package_version.isnot(None),
+                PythonPackageVersionEntity.python_package_index_id.is_(None))
+            .with_entities(
+                PythonPackageVersionEntity.package_name,
+                PythonPackageVersionEntity.package_version,
+                PythonPackageVersionEntity.python_package_index_id,
+                func.count(
+                    tuple_(
+                        PythonPackageVersionEntity.package_name,
+                        PythonPackageVersionEntity.package_version,
+                        PythonPackageVersionEntity.python_package_index_id)
+                        ))
+            .group_by(
+                PythonPackageVersionEntity.package_name,
+                PythonPackageVersionEntity.package_version,
+                PythonPackageVersionEntity.python_package_index_id)
+        )
+
+        if package_name:
+            case_2 = case_2.filter(PythonPackageVersionEntity.package_name == package_name)
+
+        if package_version:
+            case_2 = case_2.filter(PythonPackageVersionEntity.package_version == package_version)
+
+        if distinct:
+            case_2 = case_2.distinct()
+
+        if not is_count:
+            case_2 = case_2.limit(count)
+
+        result_2 = case_2.all()
+
+        case_3 = (
+            self._session.query(PythonPackageVersionEntity)
+            .filter(
+                PythonPackageVersionEntity.package_version.is_(None),
+                PythonPackageVersionEntity.python_package_index_id.is_(None))
+            .with_entities(
+                PythonPackageVersionEntity.package_name,
+                PythonPackageVersionEntity.package_version,
+                PythonPackageVersionEntity.python_package_index_id,
+                func.count(
+                    tuple_(
+                        PythonPackageVersionEntity.package_name,
+                        PythonPackageVersionEntity.package_version,
+                        PythonPackageVersionEntity.python_package_index_id)
+                        ))
+            .group_by(
+                PythonPackageVersionEntity.package_name,
+                PythonPackageVersionEntity.package_version,
+                PythonPackageVersionEntity.python_package_index_id)
+        )
+
+        if package_name:
+            case_3 = case_3.filter(PythonPackageVersionEntity.package_name == package_name)
+
+        if package_version:
+            case_3 = case_3.filter(PythonPackageVersionEntity.package_version == package_version)
+
+        if distinct:
+            case_3 = case_3.distinct()
+
+        if not is_count:
+            case_3 = []
```

Comment:
Why is this needed?

Reply:
not needed indeed :) nice catch! Thanks :D